aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+#include <ATen/FunctionalTensorImplBase.h>
+
+#include <ATen/core/List.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+
+FunctionalTensorImplBase::FunctionalTensorImplBase(c10::DispatchKeySet keyset, caffe2::TypeMeta dtype, c10::Device device)
+  : TensorImpl(
+      keyset.add(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+FunctionalTensorImplBase::FunctionalTensorImplBase(caffe2::TypeMeta dtype, c10::Device device)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+void FunctionalTensorImplBase::mark_as_alias() {
+  is_alias_tensor_ = true;
+}
+
+void FunctionalTensorImplBase::sync_() {
+  if (is_up_to_date()) {
+    return;
+  }
+  // Apply all updates on alias_
+  alias_->SyncUpdateOperations();
+  // Reapply views to Get the viewed tensor from updated base in alias_
+  auto t = alias_->base();
+  for (auto& view_meta: view_metas_) {
+    switch (view_meta.view_type) {
+      case ViewMeta::Type::view:
+          t = t.view(view_meta.size);
+          break;
+      case ViewMeta::Type::noOp:
+          break;
+      default:
+          TORCH_CHECK(false, "Tried to run the functionalization pass on an unsupported view: ", view_meta.view_type);
+    }
+  }
+  // We want the new tensor to have separate memory from the alias.
+  // Note that the clone required for functorch, but probably unnecessary for backends
+  // like vulkan and xla, because their view operators already create fresh tensors.
+  t = t.clone();
+  // This call to replace_() doesn't go through the dispatcher.
+  // It's a virtual method implemented directly on TensorImpl subclasses.
+  replace_(t.unsafeGetTensorImpl());
+  generation_ = alias_->generation();
+}
+
+bool FunctionalTensorImplBase::is_up_to_date() const {
+  if (alias_) {
+    return generation_ == alias_->generation();
+  }
+  return true;
+}
+
+void FunctionalTensorImplBase::maybe_add_update(const Tensor& updated_val) {
+  // If the mutated tensor wasn't a view, we don't need to do anything
+  if (is_view()) {
+    alias_->add_update(updated_val.clone(), view_metas_);
+    // TODO: add this back see if anything breaks?
+    //generation_ = alias_->generation(); // self is fully up to date at this point
```

Comment:
What's going on here?

Reply:
Think I hit some segfaults when I first add this, so I pushed it off. But it looks like it's working now.

The idea is that there's a precondition that makes this okay: Before we add an update to a tensor, we had to have synced that tensor with its alias first. That means that we know the tensor is still up to date with the alias, after adding the update. I just checked and it looks like the corresponding xla code for this is [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L619) (they increment the generation, but I think this expresses intent a little more clearly).

I think looking at a codegen'd inplace kernel helps:
```
at::Tensor & add__Tensor(c10::DispatchKeySet ks, at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha) {
        // sync inputs before running the operation.
        at::functionalization::impl::maybe_sync(self);
        at::functionalization::impl::maybe_sync(other);
        at::Tensor tmp_output;
        {
            at::AutoDispatchBelowFunctionalize guard;
            tmp_output = at::redispatch::add(ks & c10::after_func_keyset, self, other, alpha);
        }
        self.replace_(tmp_output);
        // At this point, self is "ahead", and we're just telling the alias about the new update.
        // Their generations should be the same by the end of this operation
        at::functionalization::impl::maybe_add_update(self);
        return self;
}
```